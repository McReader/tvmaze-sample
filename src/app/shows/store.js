import {
	createEntityAdapter,
	createAsyncThunk,
	createSlice,
} from "@reduxjs/toolkit"

const PREFIX = "shows"

export const fetchShowMainInfo = createAsyncThunk(
	`${PREFIX}/fetchShowMainInfo`,
	async (showId, thunkAPI) => {
		const result = await thunkAPI.extra.showsService.getShowMainInformation(
			showId
		)
		return result.data
	}
)

const showsAdapter = createEntityAdapter()

export const initialState = showsAdapter.getInitialState()

export const showsSlice = createSlice({
	name: PREFIX,
	initialState,
	reducers: {
		// standard reducer logic, with auto-generated action types per reducer
	},
	extraReducers: {
		[fetchShowMainInfo.fulfilled]: (state, action) => {
			showsAdapter.addOne(state, action.payload)
		},
	},
})

export const showsSelectors = showsAdapter.getSelectors((state) => state.shows)
