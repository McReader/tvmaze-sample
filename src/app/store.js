import { configureStore } from "@reduxjs/toolkit"

import { rootReducer } from "./rootReducer"
import { appContext } from "./appContext"

export const store = configureStore({
	reducer: rootReducer,
	middleware: (getDefaultMiddleware) =>
		getDefaultMiddleware({
			thunk: {
				extraArgument: appContext,
			},
		}),
})

if (process.env.NODE_ENV === "development" && module.hot) {
	module.hot.accept("./rootReducer", () => {
		const newRootReducer = require("./rootReducer").default
		store.replaceReducer(newRootReducer)
	})
}
