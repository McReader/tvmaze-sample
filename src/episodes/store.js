import { createEntityAdapter, createAsyncThunk, createSlice } from "@reduxjs/toolkit";

const PREFIX = "episodes";

export const fetchShowEpisodeList = createAsyncThunk(`${PREFIX}/fetchShowEpisodeList`, async (showId, thunkAPI) => {
    const result = await thunkAPI.extra.showsService.getShowEpisodeList(showId);
    return result.data;
});

export const fetchEpisodeMainInformation = createAsyncThunk(`${PREFIX}/fetchEpisodeMainInformation`, async ({ showId, episodeId }, thunkAPI) => {
    const episode = episodesSelectors.selectById(thunkAPI.getState(), episodeId);

    if (episode) {
        return Promise.resolve(episode);
    }

    const result = await thunkAPI.extra.episodesService.getEpisodeMainInformation(episodeId);
    return result.data;
});

const episodesAdapter = createEntityAdapter();

export const initialState = {
    ...episodesAdapter.getInitialState(),
    byShow: {},
};

export const episodesSlice = createSlice({
    name: PREFIX,
    initialState: initialState,
    reducers: {
        // standard reducer logic, with auto-generated action types per reducer
    },
    extraReducers: {
        [fetchShowEpisodeList.fulfilled]: (state, { meta, payload }) => {
            episodesAdapter.addMany(state, payload);
            state.byShow[meta.arg] = payload.map(episode => episode.id);
        },
        [fetchEpisodeMainInformation.fulfilled]: (state, { meta, payload }) => {
            episodesAdapter.addOne(state, payload);
            state.byShow[meta.arg.showId] = [payload.id];
        },
    }
});

export const episodesSelectors = {
    ...episodesAdapter.getSelectors((state) => state.episodes),
    selectByShowId: (state) => {  },
};
