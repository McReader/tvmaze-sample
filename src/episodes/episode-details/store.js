import { createSlice } from "@reduxjs/toolkit";

import { fetchEpisodeMainInformation } from "../store";

const PREFIX = "episodeDetailsPage";

const initialState = {
    requestStatus: "idle",
};

const updateRequestStatus = (state, action) => {
    state.requestStatus = action.meta.requestStatus;
};

export const episodeDetailsPageSlice = createSlice({
    name: PREFIX,
    initialState: initialState,
    reducers: {
        // standard reducer logic, with auto-generated action types per reducer
    },
    extraReducers: {
        [fetchEpisodeMainInformation.pending]: updateRequestStatus,
        [fetchEpisodeMainInformation.fulfilled]: updateRequestStatus,
        [fetchEpisodeMainInformation.rejected]: updateRequestStatus,
    }
});

export const episodeDetailsPageSelectors = {
    selectEpisodesDetailsRequestStatus: (state) => state.episodeDetailsPage.requestStatus,
};
