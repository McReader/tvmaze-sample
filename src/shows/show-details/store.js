import { createSlice } from "@reduxjs/toolkit";

import { fetchShowEpisodeList } from "../../episodes/store";

import { fetchShowMainInfo } from "../store";

const PREFIX = "showDetailsPage";

const initialState = {
    showInfoRequestStatus: "idle",
    episodesRequestStatus: "idle",
};

const updateShowInfoRequestStatus = (state, action) => {
    state.showInfoRequestStatus = action.meta.requestStatus;
};

const updateEpisodesRequestStatus = (state, action) => {
    state.episodesRequestStatus = action.meta.requestStatus;
};

export const showsDetailsPageSlice = createSlice({
    name: PREFIX,
    initialState: initialState,
    reducers: {
        // standard reducer logic, with auto-generated action types per reducer
    },
    extraReducers: {
        [fetchShowMainInfo.pending]: updateShowInfoRequestStatus,
        [fetchShowMainInfo.fulfilled]: updateShowInfoRequestStatus,
        [fetchShowMainInfo.rejected]: updateShowInfoRequestStatus,
        [fetchShowEpisodeList.pending]: updateEpisodesRequestStatus,
        [fetchShowEpisodeList.fulfilled]: updateEpisodesRequestStatus,
        [fetchShowEpisodeList.rejected]: updateEpisodesRequestStatus,
    }
});

export const showDetailsPageSelectors = {
    getEpisodesRequestStatus: (state) => state.showDetailsPage.episodesRequestStatus,
    getShowInfoRequestStatus: (state) => state.showDetailsPage.showInfoRequestStatus,
};
