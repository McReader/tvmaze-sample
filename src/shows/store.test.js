import showDetails from "../tvmaze-sdk/shows/__mocks__/showDetails.json";

import { fetchShowMainInfo, initialState, showsSlice } from "./store";

let currentState;
let nextState;
let action;

beforeEach(() => {
    nextState = showsSlice.reducer(currentState, action);
});

describe(`when action is "${fetchShowMainInfo.fulfilled}"`, () => {
    beforeAll(() => {
        currentState = initialState;
        action = fetchShowMainInfo.fulfilled(showDetails);
    });

    test("should fill the store by received items", () => {
        expect(nextState).toEqual({
            ids: [showDetails.id],
            entities: {
                [showDetails.id]: showDetails,
            },
        });
    });
});
