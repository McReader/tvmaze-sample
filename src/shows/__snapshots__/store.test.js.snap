// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when action is "shows/fetchShowMainInfo/fulfilled" should fill the store by received items 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "_links": Object {
        "previousepisode": Object {
          "href": "http://api.tvmaze.com/episodes/185054",
        },
        "self": Object {
          "href": "http://api.tvmaze.com/shows/1",
        },
      },
      "externals": Object {
        "imdb": "tt1553656",
        "thetvdb": 264492,
        "tvrage": 25988,
      },
      "genres": Array [
        "Drama",
        "Science-Fiction",
        "Thriller",
      ],
      "id": 1,
      "image": Object {
        "medium": "http://static.tvmaze.com/uploads/images/medium_portrait/81/202627.jpg",
        "original": "http://static.tvmaze.com/uploads/images/original_untouched/81/202627.jpg",
      },
      "language": "English",
      "name": "Under the Dome",
      "network": Object {
        "country": Object {
          "code": "US",
          "name": "United States",
          "timezone": "America/New_York",
        },
        "id": 2,
        "name": "CBS",
      },
      "officialSite": "http://www.cbs.com/shows/under-the-dome/",
      "premiered": "2013-06-24",
      "rating": Object {
        "average": 6.5,
      },
      "runtime": 60,
      "schedule": Object {
        "days": Array [
          "Thursday",
        ],
        "time": "22:00",
      },
      "status": "Ended",
      "summary": "<p><b>Under the Dome</b> is the story of a small town that is suddenly and inexplicably sealed off from the rest of the world by an enormous transparent dome. The town's inhabitants must deal with surviving the post-apocalyptic conditions while searching for answers about the dome, where it came from and if and when it will go away.</p>",
      "type": "Scripted",
      "updated": 1573667713,
      "url": "http://www.tvmaze.com/shows/1/under-the-dome",
      "webChannel": null,
      "weight": 97,
    },
  },
  "ids": Array [
    1,
  ],
}
`;
